### PR TITLE
builtins: preparation for having the number types derive from the numbers ABCs 

### PR DESCRIPTION
**Pull request by bluetech**
~~This should make e.g. this work:~~ (See comment https://github.com/python/typeshed/pull/3108#issuecomment-509158766).

```py
from decimal import Decimal
from numbers import Real

x: Real = 1.2
x = Decimal('1.2')
if isinstance(x, Real):
    # Not dead to mypy
```

This PR has some preparation commits before making the change. Should be easier to review each one separately.

### CODE REVIEW

**Comment by bluetech** [comment]
Hmm, this causes mypy with the new semantic analyzer to crash:

> RuntimeError: Internal error: unresolved placeholder type builtins.int

**Comment by srittau** [comment]
I like the first three commits. I am a bit wary of deriving the built-in number types from the ABCs, though. The real classes are not derived from the ABCs and probably use `ABCMeta.register()` or a similar mechanism. While I'm not totally opposed to this for practicability reasons and since this about Python's built-in numeric tower, I'd prefer to see another, more general solution, either involving protocols or some kind of typing support to registering ABCs.

**Comment by bluetech** [comment]
@srittau Thanks for taking a look, I'll remove the last commit from this PR and open an issue to discuss it.

**Comment by bluetech** [comment]
@srittau Updated. Will open the issue a bit later.

File: stdlib/2/__builtin__.pyi [hunk 1]
```diff
@@ -242,11 +245,12 @@ class float:
...
         @overload
-        def __round__(self) -> int: ...
-        @overload
-        def __round__(self, ndigits: None) -> int: ...
+        def __round__(self, ndigits: None = ...) -> int: ...
         @overload
         def __round__(self, ndigits: int) -> float: ...
+        def __trunc__(self) -> int: ...
+        def __floor__(self) -> int: ...
+        def __ceil__(self) -> int: ...
```
Comment (srittau) [inline stdlib/2/__builtin__.pyi]: The last two don't seem to exist (which surprised me):

```python
Python 3.7.3 (default, Jun  3 2019, 18:27:33) 
[GCC 6.3.0 20170516] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> x = 3.0
>>> x.__trunc__()
3
>>> x.__floor__()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'float' object has no attribute '__floor__'
>>> x.__ceil__()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'float' object has no attribute '__ceil__'
```

Comment (bluetech) [reply stdlib/2/__builtin__.pyi]: Hmm, should have checked all of them.

I looked into cpython code, and what I understand is that `math.floor` (same for `math.ceil`) has a special case for floats, or rather types that can be converted to float (implement `__float__()`). In these cases, if `__floor__` is not defined, it converts to float and does the operation directly.

```py
import math

class X:
    def __float__(self):
        return 5.4

    #def __floor__(self):
    #    return 10

print(math.floor(X()))
```

Commented it prints 5, uncommented it prints 10.

So I think that, given `numbers.Real` already requires `__float__`, we can just diverge from the actual definition of `numbers.Real` and remove `__floor__` and `__ceil__` from it in typeshed. WDYT? [The `numbers.Real` doc](https://docs.python.org/3/library/numbers.html#numbers.Real) only requires `math.floor` to work.

File: stdlib/2and3/numbers.pyi [hunk 2]
```diff
@@ -70,7 +70,11 @@ class Real(Complex, SupportsFloat):
         @abstractmethod
         def __ceil__(self) -> int: ...
         @abstractmethod
-        def __round__(self, ndigits: Optional[int] = ...): ...
+        @overload
+        def __round__(self, ndigits: None = ...): ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/numbers.pyi]: Not sure why this needs overloads.

File: stdlib/2/__builtin__.pyi [hunk 3]
```diff
@@ -168,7 +168,7 @@ class int:
     def __rtruediv__(self, x: int) -> float: ...
     def __rmod__(self, x: int) -> int: ...
     def __rdivmod__(self, x: int) -> Tuple[int, int]: ...
-    def __pow__(self, x: int) -> Any: ...  # Return type can be int or float, depending on x.
+    def __pow__(self, x: int, modulo: Optional[int] = ...) -> Any: ...  # Return type can be int or float, depending on x.
```
Comment (JelleZijlstra) [inline stdlib/2/__builtin__.pyi]: The arguments are actually positional-only

Comment (bluetech) [reply stdlib/2/__builtin__.pyi]: Is there some special way to encode this?

File: stdlib/2and3/numbers.pyi (hunk 2 above)
Comment (bluetech) [reply stdlib/2and3/numbers.pyi]: The idea is to have the same form as `float` above. Once this stub is filled the return types of the two overloads will be different, I think.

File: stdlib/2/__builtin__.pyi (hunk 3 above)
Comment (JelleZijlstra) [reply stdlib/2/__builtin__.pyi]: Start the names with two underscores, like `__modulo`.

**Comment by bluetech** [comment]
I stripped the useful parts of this PR to separate PRs and opened an issue about the `__ceil__`/`__floor__` situation.

**Pull request closed without merging**